### PR TITLE
Exponential Auto-Reconnect

### PR DESCRIPTION
**Pull request by HJianBo**
As discussed in #286 and PR #287

### CODE REVIEW

**Comment by HJianBo** [comment]
@sachinvas @Abhijith-net @halonsoluis  Please review it

File: Source/CocoaMQTT.swift [hunk 1]
```diff
@@ -207,8 +207,20 @@ public class CocoaMQTT: NSObject, CocoaMQTTClient, CocoaMQTTDeliverProtocol {
...
-    /// Auto reconnect time interval
-    public var autoReconnectTimeInterval: UInt16 = 20
+    /// Reconnect time interval
+    ///
+    /// - note: This value will be increased with `autoReconnectTimeInterval *= 2`
+    ///         if reconnect failed
+    public var autoReconnectTimeInterval: UInt16 = 1 // starts from 1 second
+    
+    private var reconnectExponetials: Array<UInt16> = []
+    private var reconectCheckTimes = 0
```
Comment (Abhijith-net) [inline Source/CocoaMQTT.swift]: @HJianBo `reconectCheckTimes` typo, should be `reconnectCheckTimes`

File: Source/CocoaMQTT.swift [hunk 2]
```diff
@@ -207,8 +207,20 @@ public class CocoaMQTT: NSObject, CocoaMQTTClient, CocoaMQTTDeliverProtocol {
     /// Enable auto-reconnect mechanism
     public var autoReconnect = false
     
-    /// Auto reconnect time interval
-    public var autoReconnectTimeInterval: UInt16 = 20
+    /// Reconnect time interval
+    ///
+    /// - note: This value will be increased with `autoReconnectTimeInterval *= 2`
+    ///         if reconnect failed
+    public var autoReconnectTimeInterval: UInt16 = 1 // starts from 1 second
```
Comment (Abhijith-net) [inline Source/CocoaMQTT.swift]: @HJianBo Ideally, would like to have a separate class for reconnect functionality. 
Which shall have 
init() -> which initializes the array of timeouts
reconnectIfRequired() -> which shall check whats the next timer and starts the timer. If we have reached the maxReconnectTime, we reset
reset() -> reset the properties like 
                    self.reconnectExponetials = []
                    self.reconectCheckTimes = 0
                    self.autoReconnTimer = nil

Comment (HJianBo) [reply Source/CocoaMQTT.swift]: Yes, it a good idea to improve readability and usability of code. I'll fix later  : ]

File: Source/CocoaMQTT.swift [hunk 3]
```diff
@@ -490,18 +502,45 @@ extension CocoaMQTT: GCDAsyncSocketDelegate {
...
-        autoReconnTimer = nil
+        
         if disconnectExpectedly {
             connState = .initial
-        } else if autoReconnect && autoReconnectTimeInterval > 0 {
+        } else if autoReconnect {
+            
+            // Generate exponential time point array
+            var exponential = autoReconnectTimeInterval
+            while(true) {
```
Comment (sachinvas) [inline Source/CocoaMQTT.swift]: Hi @HJianBo, the disconnect gets called on every unsuccessful retry and the exponential array is flooded with multiple repetitions of same set of values `[1 to maxAutoReconnectTimeInterval]`.

Comment (HJianBo) [reply Source/CocoaMQTT.swift]: Yep, it's a bug. The exponential array should be generated once time only.  I'll fixed it later

**Comment by sachinvas** [comment]
@HJianBo, Is there any specific reason to create the array of exponential retry values? The potential pitfall with this is that we don't reconnect at all after we remove the exponential retry values.

I didn't understand what is the design error in the previous PR?

**Comment by HJianBo** [comment]
@sachinvas 

In the previous PR, It just update the `autoReconnectTimeInterval` value in the every Timer checking event. But, It doesn't update the checking interval for timer actually.

So, In the current  mechanism. We use `reconnectExponetials` variable to record a list of time points. It indicate to the reconnection action executed position in the cycle checking

**Comment by sachinvas** [comment]
@HJianBo
In the previous PR, the design is as below:
1. Start the reconnect timer at the autoReconnectTimeInterval.
2. Inside the reconnect timer update the autoReconnectTimeInterval to next time point.
3. Once we reach the maxAutoReconnectTimeInterval, use that value for subsequent reconnection timer.
4. So, ideally we reach maxAutoReconnectTimeInterval seconds at the end of only one cycle and use the same value (maxAutoReconnectTimeInterval) unless we connect again.
I am seeing the battery impact is bit high with executing the same list of time points for every cycle. This is even addressed in the above design.
Please let me know your thoughts.

**Comment by HJianBo** [comment]
Yes @sachinvas You are right. I missed the old CocoaMQTTTimer instance will be destructed after giving the `autoReconnTimer` a New value.

Besides, I think it is important to make `autoReconnectTimeInterval` as public, not give it a constant value 1.

I'll restore your changes and fix a little problems to prepare release 1.2.5  :)

**Comment by sachinvas** [comment]
Hi @HJianBo, thank you so much for your consideration.

I am also facing `com.apple.root.default-qos.overcommit` issue. Looks like this is because of https://forums.developer.apple.com/thread/110523. We need to create a private queue to handle the `CocaMQTTTimer` eventHandler.

Please do let me know your thoughts.

**Comment by HJianBo** [comment]
@sachinvas Cool!  Thanks for your feedback it!

In the current version, there are many operations that need to support asynchronous queues, e.g: timer, network I/O. We plan to implement it at v1.3.0

Anyway, we should release this reconnect feature first

**Comment by Abhijith-net** [comment]
Thanks, @HJianBo, and @sachinvas for the fix!

**Pull request merged**